File: Go/106-construct-binary-tree-from-inorder-and-postorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func buildTree(inorder []int, postorder []int) *TreeNode {
    if len(inorder) == 0 {
		return nil
	}

	root := new(TreeNode)
    root.Val = postorder[len(postorder) - 1]
	rootIdx := 0
	for inorder[rootIdx] != root.Val {
		rootIdx++
	}
    root.Left = buildTree(inorder[:rootIdx], postorder[:rootIdx])
    root.Right = buildTree(inorder[rootIdx + 1:], postorder[rootIdx :len(postorder) - 1])
	return root
}

File: Python/106-construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if inorder == []:
            return None
        root = TreeNode(postorder.pop()) #根节点
        index = inorder.index(root.val) #在中序结果中找到根节点的位置
        #递归构造左右子树
        root.left = self.buildTree(inorder[:index], postorder[:index])
        root.right = self.buildTree(inorder[index + 1:], postorder[index:])
        return root